### DISCUSSION TITLE
Communication between Machines and how to extend.

### DISCUSSION THREAD

**Discussion by ghost** [statelyai/xstate#1894] [Q&A]
Hey People quick forewarning completly new to state machines and such: 

Question 1:
 
Let's say I have a Graph of X nodes and Y edges, A node can be Visible( with substates) or hidden and has an Id in its context plus additional Information. 
An edge is visible or hidden and has Target and Source in its context.
Now I have multiple filters that would change the state of one or more Nodes only the Nodes that match the filter arguments would react to the event thanks to a guard. 

So what I want is for a node to tell the list of edges that it got hidden. 
Making all Edges that have the NodeId in either Target or Source transition themself to hidden. This could be done with a guard on the event as far as I understood. 
But how do I tell all edge Machines about the event.

Secondly if booth nodes are Hidden and one Node gets visible again.
The edge should update its context but stay hidden and only change its own state when booth Nodes get visible again.
Could machines or services talk to each other in that way? 
Specialy if the amount of Nodes and Edges is dynamic since they can be Loaded at a later time. When requested by the user.

Question 2: 

Lets say I have a Machine that has two states, Visible and Hidden. 
That is equal for all my components. 

But visible can have diffrent Events or even substates based on the Component. 
So would their be a way to extend a Machine / state?

**Comment by ezhikov** [comment]
I'm not an expert on communication between machines, so I can talk only about basic and pretty abstract stuff, but I think it might help.

All your machines communicate with each other and with themselves via Actor Model. Basis of Actor model is that each machine (and actually all other stuff) is an Actor. Actor can send messages to another actor it knows. For example, if you spawning machine, your spawned machines automatically know about their parent and you keep reference to child machines in context. You can also send references to other actors you know about. You can also set up some kind of message bus (which will also be an actor, but very chatty one). There is numerous ways to implement this kind of communication and without actual code I can't even recommend something useful.

In your particular situation you need each Edge to _known_ about its Nodes and for your Nodes to know about its Edges. When Node sends "I'm hidden" message to Edge, Edge notes in it's internal state which Node is now hidden and also goes into hidden state. When this Node reappears, edge checks if both Nodes are visible and transition to "visible" state it they are. 

As for configuration, check out this part of the documentation: [Extending machines](https://xstate.js.org/docs/guides/machines.html#extending-machines)

**Reply by tomByrer** [reply]
Maybe the actor model would be best here!

But for some reason, I didn't use one, but a Master with Sub-Machines that are in an array.  If I needed to trigger all, then I just looped or `.map` over that array.  Also the Master & Subs can be in different environments (eg Master is on the server, & Subs in clients).

I tried to use Visible and Hidden in a parallel sub-state, but found using context is easier.  That way my Sub-Machines only have 1 state vs 2, and the context is easier to [configure outside of the machine running](https://xstate.js.org/docs/guides/machines.html#initial-context).
```js

{
  actions: {
    disable: assign({ isEnabled: false }),
    enable: assign({ isEnabled: true }),
  },
  guards: {
    canEdit: ctx => ctx.isEnabled,
  }
}

**Reply by ezhikov** [reply]
> I didn't use one, but a Master with Sub-Machines that are in an array.

This is still an actor model. Actor model is an abstraction that can be implemented differently. Some implementation have awesome features, like supervisors and recovery strategies, but it is not mandatory. Basics is that actors are single threaded, non blocking units of computation that have isolated state and can send and receive messages. How you implement communication for each particular case may be different. It can be websocket, http request, DOM event. And you can say "all actors are in this array, so send them messages directly" or you can maintain hierarchy, or you can pass references to actors around in messages. It's up to you.

When we migrated from Redux there was a moment when we had most of data fetching already in XState, but still had to store that data in Redux. So I told machine how to send message to redux and I had system of two actors - XState service that fetched data, and Redux that supplied that data to components.

> I tried to use Visible and Hidden in a parallel sub-state, but found using context is easier.

My rule of thumb is if something can be put in distinct state - it goes in state, even if that state derived from context. And I don't really like parallel states. There is a fine line after which you don't understand how next event will affect all that parallel states.
